lib/engine/game/g_system18/step/buy_sell_par_bid_merge.rb

```diff
@@ -0,0 +1,193 @@
+
+            @log << "#{entity.name} wins bid on #{corporation.name} for #{@game.format_currency(price)}"
+            par_price = price / 2
+
+            share_price = get_all_par_prices(corporation).find { |sp| sp.price <= par_price }
+
+            # Temporarily give the entity cash to buy the corporation PAR shares
+            @game.bank.spend(share_price.price * 2, entity)
+
+            action = Action::Par.new(entity, corporation: corporation, share_price: share_price)
+            process_par(action)
+
+            # Clear the corporation of 'share' cash grabbed earlier.
+            corporation.spend(corporation.cash, @game.bank)
+
+            # Then move the full amount.
+            entity.spend(price, corporation)
+
+            @auctioning = nil
+
+            # Player to the right of the winner is the new player
+            @round.goto_entity!(winner.entity)
+            pass!
+          end
+
+          def can_bid_any?(entity)
+            max_bid(entity) >= MIN_BID && !bought? &&
+            @game.corporations.any? do |c|
+              @game.can_par?(c, entity) && c.type == :minor && can_buy?(entity, c.shares.first&.to_bundle)
+            end
+          end
+
+          def can_ipo_any?(entity)
+            @game.phase.name.to_i >= MAJOR_PHASE && !bought? &&
+            @game.corporations.any? do |c|
+              @game.can_par?(c, entity) && c.type == :major && can_buy?(entity, c.shares.first&.to_bundle)
+            end
+          end
+
+          def get_all_par_prices(_corp)
+            # minors and majors can start at the same values
+            @game.stock_market.share_prices_with_types(%i[par])
+          end
+
+          def ipo_type(entity)
+            # Major's are par, minors are bid
+            phase = @game.phase.name.to_i
+            case entity.type
+            when :major
+              if phase >= MAJOR_PHASE
+                if @game.home_token_locations(entity).empty?
+                  'No home token locations are available'
+                else
+                  :par
+                end
+              else
+                "Cannot start till phase #{MAJOR_PHASE}"
+              end
+            when :national
+              'Cannot start'
+            else
+              :bid
+            end
+          end
+
+          def can_convert?(minor)
+            (minor.owner == current_entity) && (minor.type == :minor) &&
+              minor.floated? && !bought?
+          end
+
+          def can_convert_any?(_entity)
+            return false if @game.phase.name.to_i < MERGE_PHASE
+            return false if bought?
+
+            @game.corporations.any? { |corporation| can_convert?(corporation) }
+          end
+
+          def connected?(minor_a, minor_b)
+            # Mergeable candidates must be connected by track, minors only have one token which simplifies it
+            # (partially borrowed from 1867)
+            # FIXME: avoid graph by using destination_connection
```

Comment:
Any chance this could be addressed before merging?

Reply:
I've been struggling to figure how to do it at all. There are two difficulties with this. First is that merging is just one action of many in a stock round turn. This means that I can't just auto-insert a pass if a minor isn't connected to another minor. The second is that destination_connection just supplies a list of entities that have reached their destination. I really need a list of minors with each entry containing a list of minors that minor reaches. I would have to create a new action - and I'm really trying to avoid touching common code with all of these System18 maps.

In the end, I decided that the maps in System18 were small enough that graph wasn't nearly expensive as in a "big" 18xx.  Maybe I should just pull out the FIXME.

Reply:
I just realized this is even harder: when a major is started in a stock round, it can drop its home token in such a way that it disconnects formerly connected minors. This means any auto-action I generate in an OR is pointless, so any "connection" action I create would have to be a stock action.

Reply:
Hold on, I just realized something. I don't need to know whether a minor is connected in the buy_sell_shares action method! The same action - "choose" is used for both converting and merging minors. Any minor that can be merged is always guaranteed to be able to convert. As long as I don't validate merge candidates calling graph will only occur as a result of rendering. So depending on how I coded this (I'll take a look in the morning), I don't think there is a performance issue at all here with graph.